### PR TITLE
Implement Theme model/controller and Theme update page

### LINKED ISSUE #1090: Theming : Settings can be changed
The settings that make up a theme can be loaded, changed, and saved back to the server.

* As a manager I can update the application theme

  - [ ] Colors can be set
  - [ ] ~~Brand image can be set~~ Moved to next sprint.
  - [ ] Theme can be undone

Belongs to #1067

### LINKED ISSUE #1091: Theming : Settings storage
The theme settings can be saved and loaded from the server as json data.

Belongs to #1067

### PR DESCRIPTION
**Pull request by voxlol**
Closes #1090 
Closes #1091 

Extension of PR of #1107 (but also implements UX - theme storage) and includes both Allens' commits

#### Screenshots
screen shot 2015-12-23 at 2 17 10 pm

screen shot 2015-12-23 at 12 23 06 pm
screen shot 2015-12-23 at 12 23 10 pm

### CODE REVIEW

File: db/seeds.rb [hunk 1]
```diff
@@ -1,3 +1,4 @@
 # require_relative 'seeds/products'
 # require_relative 'seeds/project_questions'
 # require_relative 'seeds/staff.rb'
+# require_relative 'seeds/theme.rb'
```
Comment (stackus) [inline db/seeds.rb]: The other lines were commented out for '???' reasons. This new line won't be run while it's commented.

File: db/seeds/theme.rb [hunk 2]
```diff
@@ -0,0 +1,2 @@
+default_theme_data = YAML.load_file(File.join [Rails.root, 'db', 'data', 'sample', 'themes.yml'])
```
Comment (stackus) [inline db/seeds/theme.rb]: Given we're working to break demo/sample data out of the core application we should avoid putting seed data into the sample directory.

File: spec/factories/projects.rb [hunk 3]
```diff
@@ -37,7 +37,7 @@
     spent 0.0
     monthly_spend 0
     start_date((Time.zone.now + 1.week).to_date)
-    end_date((Time.zone.now + 2.week).to_date)
+    end_date((Time.zone.now + 2.weeks).to_date)
```
Comment (stackus) [inline spec/factories/projects.rb]: While `weeks` reads better it makes no difference to the code which one is used.

Try to avoid making changes to unrelated code.

File: spec/requests/theme_spec.rb [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  describe 'GET show' do
+    it 'returns the theme for an admin' do
+      sign_in_as create(:staff, :admin)
+
+      get theme_path
+
+      expect(response.body).to eq(Theme.first.to_json)
+    end
+
+    it 'returns a the theme for a user' do
```
Comment (stackus) [inline spec/requests/theme_spec.rb]: Typo

Here and on the next test.

File: spec/requests/theme_spec.rb [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  end
+
+  describe 'PUT update' do
+    before(:each) do
+      @theme = create :theme
+      @params = {
+        name: 'Theme name',
+        description: 'Theme description',
+        config: {
+          global: ['mock'],
```
Comment (stackus) [inline spec/requests/theme_spec.rb]: This doesn't look like valid input. I'd expect this data to fail in the "allows an admin to make updates" test.

File: src/client/app/components/forms/forms/theme.form.js [hunk 6]
```diff
@@ -0,0 +1,107 @@
+(function() {
+  'use strict';
+
+  angular.module('app.components')
+    .config(themeForm);
+
+  /** @ngInject */
+  function themeForm(FormsProvider) {
+    FormsProvider.register('themeForm', {
+      fields: [
```
Comment (stackus) [inline src/client/app/components/forms/forms/theme.form.js]: This is full of repetition.

File: src/client/app/resources/theme.factory.js [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+      update: {
+        method: 'PUT',
+        isArray: false
+      },
+      query: {
+        isArray: false
+      }
+    });
+
+    theme.defaults = {
```
Comment (stackus) [inline src/client/app/resources/theme.factory.js]: This is the third time I've seen the exact same list. The theme implementation needs to work without this structure being repeated everywhere.

File: src/client/index.html [hunk 8]
```diff
@@ -73,6 +74,9 @@
   <script src="/bower_components/marked/lib/marked.js"></script>
   <script src="/bower_components/angular-marked/angular-marked.js"></script>
   <script src="/bower_components/angular-aria/angular-aria.js"></script>
+  <script src="/bower_components/tinycolor/tinycolor.js"></script>
+  <script src="/bower_components/bootstrap/dist/js/bootstrap.js"></script>
```
Comment (stackus) [inline src/client/index.html]: Something is bringing in bootstrap.js. We're using ui.bootstrap an Angular-ized version and this shouldn't be needed.

File: spec/factories/projects.rb (hunk 3 above)
Comment (voxlol) [inline spec/factories/projects.rb]: Rubocop pointed out the error, should I ignore?

Comment (stackus) [inline spec/factories/projects.rb]: Odd that it decided to care now. Must be a recent update to rubucop. I didn't check the commit logs (terrible on my part) but if you make a change like this put it into it's own commit with a message saying why.

Leave it in. Have to keep our Rubocop taskmaster happy.

File: src/client/app/components/forms/forms/theme.form.js (hunk 6 above)
Comment (voxlol) [inline src/client/app/components/forms/forms/theme.form.js]: Should be addressed in latest commit

File: db/seeds/themes.yml [hunk 9]
```diff
@@ -0,0 +1,229 @@
+---
+- name: Default
+  description: Default theme for Project Jellyfish
+  config: {
```
Comment (stackus) [inline db/seeds/themes.yml]: Does this parse correctly? I'd expect no brackets or commas.

```yaml
config:
  global:
    - type: style
      label: Primary Background
      selector: primary-background
      rule: background-color
      value: "#EBEBEB"
    - type: style
      label: ...
```

File: src/client/app/components/forms/fields.config.js [hunk 10]
```diff
@@ -69,5 +69,31 @@
...
+        label: 'Color',
+        colorPickerFormat: '"hex"',
+        colorPickerAlpha: false,
+        colorPickerPos: '"top left"',
+        colorPickerSwatchBootstrap: false
+      }
+    });
+
+    // Generic color mixin
+    Forms.fields('mixin', {
```
Comment (stackus) [inline src/client/app/components/forms/fields.config.js]: This matches the above field exactly.

File: src/client/app/components/forms/formly.config.js [hunk 11]
```diff
@@ -483,6 +520,87 @@
         }
       }
     }
+
+    function multipleColorsField() {
```
Comment (stackus) [inline src/client/app/components/forms/formly.config.js]: This is more or less a copy-paste from another field. I know this because there's code in this field and associated controller that doesn't belong.

@AllenBW @voxlol  Please go over and grok the code and remove the unnecessary bits.

**Comment by voxlol** [comment]
@stackus I merged the two nested form controllers into one...but it felt a little bit hacky because I had to add in some ternaries and configuration options. Is there a simpler approach to refactoring it?

File: src/client/app/components/forms/formly.config.js [hunk 12]
```diff
@@ -318,91 +346,62 @@
...
+    multipleOptionsField();
 
     function questionsField() {
-      formlyConfig.setType({
+      formlyConfig.setType(angular.extend({
         name: 'questions',
-        template: '<formly-form form="form" model="options.data.values" ' +
-          'fields="options.data.fields" options="formOptions"></formly-form>',
-        defaultOptions: {
+        controller: nestedFormCtrlFactory({
```
Comment (stackus) [inline src/client/app/components/forms/formly.config.js]: `Ctrl` is not convention.

File: src/client/app/components/forms/formly.config.js [hunk 13]
```diff
@@ -318,91 +346,62 @@
...
+    function multipleColorsField() {
+      formlyConfig.setType(angular.extend({
+        name: 'multipleColors',
+        controller: nestedFormCtrlFactory({
+          templateOptions: ['label'],
+          data: {
+            indexKey: 'selector',       // Key to index all the different data
+            location: 'config',         // Key to data from resource
+            buildKey: 'type',           // Key to build field from field data
+            transform: mapMixinToStyle  // Change which form field to use
```
Comment (stackus) [inline src/client/app/components/forms/formly.config.js]: This was the source of my original comment. Why is the color field being hinged on the `type` value? Both have the exact same output, a colorpicker.

Because the original nested color picker code was copied without modification you've got this unnecessary requirement on the `type`.

File: src/client/app/components/forms/formly.config.js [hunk 14]
```diff
@@ -318,91 +346,62 @@
...
-
-            if (newValue === oldValue) {
-              return;
-            }
-
-            question.value = newValue;
-          }
-        }
+      function mapMixinToStyle(fieldType) {
+        return fieldType === 'mixin' ? 'style' : fieldType;
```
Comment (stackus) [inline src/client/app/components/forms/formly.config.js]: Why not just `return 'style';`?

File: src/client/app/components/forms/formly.config.js [hunk 15]
```diff
@@ -483,6 +482,85 @@
...
+      var templateOptions = config.templateOptions;
+      var buildKey = config.data.buildKey;
+      var indexKey = config.data.indexKey;
+      var location = config.data.location;
+      var transform = config.data.transform || null;
+
+      return nestedFormCtrl;
+
+      /** @ngInject */
+      function nestedFormCtrl($scope, lodash, Forms) {
```
Comment (stackus) [inline src/client/app/components/forms/formly.config.js]: `Ctrl` is not convention.

File: src/client/app/components/forms/fields.config.js [hunk 16]
```diff
@@ -69,5 +69,18 @@
         label: 'Email'
       }
     });
+
+    // Generic color
+    Forms.fields('style', {
```
Comment (stackus) [inline src/client/app/components/forms/fields.config.js]: This should be `colorpicker`.

**Comment by stackus** [comment]
:+1:
